### PR TITLE
@pydef improvements

### PR DESCRIPTION
**Pull request by cstjean**
I used [`MacroTools.splitdef`](https://github.com/MikeInnes/MacroTools.jl#function-definitions) to clean up the method definition code. This enables support for long-form function definitions (along with parametric types and return types, which might not get much use). I also now create types with a method dictionary. It will help #470.

(Sorry for the other PR; I had forgotten to branch)

### CODE REVIEW

**Comment by stevengj** [comment]
What are all the 0.6 test failures?

**Comment by cstjean** [comment]
Fixed.

File: README.md [hunk 1]
```diff
@@ -409,10 +409,14 @@ For instance,
...
-        __init__(self, x=10) = (self[:x] = x)
+        function __init__(self, x=10)
+	    self[:x] = x
+	end
         my_method(self, arg1::Number) = arg1 + 20
         x2.get(self) = self[:x] * 2
-        x2.set!(self, new_val) = (self[:x] = new_val / 2)
+        function x2.set!(self, new_val)
+	    self[:x] = new_val / 2
+	end
```
Comment (stevengj) [inline README.md]: indent

Comment (stevengj) [reply README.md]: What's wrong with the one-liner definitions?

**Comment by cstjean** [comment]
Fixed indentation.

**Comment by rdeits** [comment]
I think this is causing errors on v0.7 that are being masked by the fact that the 0.7 build was stalling out. Previously, the 0.7 build would hit https://github.com/JuliaLang/julia/issues/26357 but now it errors out earlier: https://travis-ci.org/JuliaPy/PyCall.jl/jobs/357206714#L674

As I understand it, that means that the `with_ioraise` macro is being called with 0 arguments instead of the 1 it's expecting. It's not clear how that could be happening, but I guess macro parsing inside other macros is different in 0.7?

**Comment by cstjean** [comment]
Looks like some interaction with https://github.com/JuliaLang/julia/pull/21746 ?

> All macros receive an extra argument __source__::LineNumberNode which describes the parser location in the source file for the @ of the macro call. It can be accessed as a normal argument variable in the body of the macro. This is implemented by inserting an extra leading argument into the Expr(:macrocall, :@name, LineNumberNode(...), args...) surface syntax.

It shouldn't affect us, though. Maybe it's a bug?

[This line](https://github.com/JuliaPy/PyCall.jl/blob/55648f27682835cbe886b9c6de39a25ad8c8c1bd/src/exception.jl#L68) may have to be changed, but it looks unrelated.

**Comment by rdeits** [comment]
I don't think it's the new line number argument. The error message is: 

```
ERROR: LoadError: LoadError: LoadError: MethodError: no method matching @with_ioraise(::LineNumberNode, ::Module)
Closest candidates are:
  @with_ioraise(::LineNumberNode, ::Module, ::Any) at /home/travis/.julia/v0.7/PyCall/src/io.jl:22
```

which shows that the macro was defined to take the `LineNumberNode`, the `Module`, and then it's "actual" argument `expr::Any`, but was called with just the `LineNumberNode` and `Module`. If it were an issue of the missing LineNumberNode, then the first argument types wouldn't match.

**Comment by Keno** [comment]
What's happening is that pydef now generates definitions like this:
```
    $(Expr(:escape, :(function ##fileno#1531(self; )::Any
      #= /home/keno/.julia/v0.7/MacroTools/src/utils.jl:260 =#
      begin
          @with_ioraise
      end
```

**Comment by Keno** [comment]
I'll have a fix shortly.

**Comment by Keno** [comment]
See https://github.com/MikeInnes/MacroTools.jl/pull/68.

**Pull request merged**